File: src/domain/dynadot.ts
import type { Currency, SearchResult } from "@/shared/types";

const DYNADOT_API_URL = "https://api.dynadot.com/api3.json";

type DynadotSearchRow = {
  DomainName?: string;
  Available?: string;
  Price?: string;
  Status?: string;
};

type DynadotSearchResponse = {
  SearchResponse?: {
    ResponseCode?: string;
    Error?: string;
    SearchResults?: DynadotSearchRow[];
  };
  Response?: {
    ResponseCode?: string;
    Error?: string;
  };
};

const timeoutSignal = (timeoutMs: number): AbortSignal => {
  const controller = new AbortController();
  setTimeout(() => controller.abort(), timeoutMs);

  return controller.signal;
};

export const parseDynadotPrice = (value: string | undefined): string | null => {
  if (!value) return null;

  const registrationMatch = value.match(/Registration Price:\s*([0-9]+(?:\.[0-9]+)?)/i);
  if (registrationMatch?.[1]) return registrationMatch[1];

  const genericMatch = value.match(/([0-9]+(?:\.[0-9]+)?)\s+in\s+USD/i);

  return genericMatch?.[1] ?? null;
};

export const affiliateUrl = (domain: string, template?: string): string => {
  const fallback = `https://www.dynadot.com/domain/search?domain=${encodeURIComponent(domain)}`;
  if (!template || template.trim().length === 0) return fallback;

  if (template.includes("{domain}")) {
    return template.replaceAll("{domain}", encodeURIComponent(domain));
  }

  try {
    const url = new URL(template);
    url.searchParams.set("domain", domain);

    return url.toString();
  } catch {
    return fallback;
  }
};

const requestOne = async (params: {
  apiKey: string;
  domain: string;
  currency: Currency;
  timeoutMs: number;
}): Promise<DynadotSearchResponse> => {
  const query = new URLSearchParams({
    key: params.apiKey,
    command: "search",
    show_price: "1",
    currency: params.currency,
    domain0: params.domain,
  });

  try {
    const response = await fetch(`${DYNADOT_API_URL}?${query.toString()}`, {
      method: "GET",
      signal: timeoutSignal(params.timeoutMs),
    });

    return (await response.json()) as DynadotSearchResponse;
  } catch {
    throw new Error("Dynadot request timed out");
  }
};

const mapResult = (params: {
  domain: string;
  payload: DynadotSearchResponse;
  currency: Currency;
  affiliateTemplate: string | undefined;
}): SearchResult => {
  const rows = params.payload.SearchResponse?.SearchResults ?? [];
  const row = rows.find((item) => item.DomainName?.toLowerCase() === params.domain) ?? rows[0];
  const available = row?.Available?.toLowerCase() === "yes";

  return {
    domain: params.domain,
    available,
    price: available ? parseDynadotPrice(row?.Price) : null,
    currency: params.currency,
    buyUrl: available ? affiliateUrl(params.domain, params.affiliateTemplate) : null,
    source: "dynadot",
    cached: false,
    quotedAt: new Date().toISOString(),
    ...(row?.Status && row.Status !== "success" ? { error: row.Status } : {}),
  };
};

const ensureSuccess = (payload: DynadotSearchResponse): void => {
  const authCode = payload.Response?.ResponseCode;
  if (authCode === "-1") {
    const message = payload.Response?.Error ?? "Dynadot authentication failed";
    if (message.toLowerCase().includes("invalid key")) {
      throw new Error("Invalid Dynadot key.");
    }

    throw new Error(message);
  }

  const searchCode = payload.SearchResponse?.ResponseCode;
  if (searchCode && searchCode !== "0") {
    const message = payload.SearchResponse?.Error ?? "Dynadot search failed";
    if (message.toLowerCase().includes("invalid key")) {
      throw new Error("Invalid Dynadot key.");
    }

    throw new Error(message);
  }
};

export const searchDynadot = async (params: {
  apiKey: string;
  domains: string[];
  currency: Currency;
  timeoutMs: number;
  affiliateTemplate: string | undefined;
}): Promise<SearchResult[]> => {
  const results: SearchResult[] = [];

  for (const domain of params.domains) {
    const payload = await requestOne({
      apiKey: params.apiKey,
      domain,
      currency: params.currency,
      timeoutMs: params.timeoutMs,
    });
    ensureSuccess(payload);
    results.push(
      mapResult({
        domain,
        payload,
        currency: params.currency,
        affiliateTemplate: params.affiliateTemplate,
      }),
    );
  }

  return results;
};

File: internal/domain/dynadot.go
// Package domain implements domain registrar API clients.
package domain

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"dotld/internal/shared"
)

const dynadotAPIURL = "https://api.dynadot.com/api3.json"

type dynadotSearchRow struct {
	DomainName string `json:"DomainName"`
	Available  string `json:"Available"`
	Price      string `json:"Price"`
	Status     string `json:"Status"`
}

type dynadotSearchResults struct {
	ResponseCode  string             `json:"ResponseCode"`
	Error         string             `json:"Error"`
	SearchResults []dynadotSearchRow `json:"SearchResults"`
}

type dynadotResponse struct {
	ResponseCode string `json:"ResponseCode"`
	Error        string `json:"Error"`
}

type dynadotSearchResponse struct {
	SearchResponse *dynadotSearchResults `json:"SearchResponse"`
	Response       *dynadotResponse      `json:"Response"`
}

var (
	registrationPriceRe = regexp.MustCompile(`(?i)Registration Price:\s*([0-9]+(?:\.[0-9]+)?)`)
	genericPriceRe      = regexp.MustCompile(`(?i)([0-9]+(?:\.[0-9]+)?)\s+in\s+USD`)
)

func ParseDynadotPrice(value string) *string {
	if value == "" {
		return nil
	}

	m := registrationPriceRe.FindStringSubmatch(value)
	if len(m) > 1 {
		return &m[1]
	}

	m = genericPriceRe.FindStringSubmatch(value)
	if len(m) > 1 {
		return &m[1]
	}

	return nil
}

func AffiliateURL(domain string, template string) string {
	fallback := "https://www.dynadot.com/domain/search?domain=" + url.QueryEscape(domain) + "&rscreg=github"
	if template == "" || strings.TrimSpace(template) == "" {
		return fallback
	}

	if strings.Contains(template, "{domain}") {
		return strings.ReplaceAll(template, "{domain}", url.QueryEscape(domain))
	}

	u, err := url.Parse(template)
	if err != nil {
		return fallback
	}
	q := u.Query()
	q.Set("domain", domain)
	u.RawQuery = q.Encode()

	return u.String()
}

func requestOne(ctx context.Context, apiKey, domain, currency string) (*dynadotSearchResponse, error) {
	q := url.Values{
		"key":        {apiKey},
		"command":    {"search"},
		"show_price": {"1"},
		"currency":   {currency},
		"domain0":    {domain},
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, dynadotAPIURL+"?"+q.Encode(), nil)
	if err != nil {
		return nil, err
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, errors.New("Dynadot request timed out")
	}
	defer resp.Body.Close()

	var result dynadotSearchResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to parse Dynadot response: %w", err)
	}

	return &result, nil
}

func mapResult(domain string, payload *dynadotSearchResponse, currency, affiliateTemplate string) shared.SearchResult {
	var rows []dynadotSearchRow
	if payload.SearchResponse != nil {
		rows = payload.SearchResponse.SearchResults
	}

	var row *dynadotSearchRow
	for i := range rows {
		if strings.ToLower(rows[i].DomainName) == domain {
			row = &rows[i]
			break
		}
	}
	if row == nil && len(rows) > 0 {
		row = &rows[0]
	}

	available := row != nil && strings.ToLower(row.Available) == "yes"

	result := shared.SearchResult{
		Domain:    domain,
		Available: available,
		Currency:  currency,
		Cached:    false,
		QuotedAt:  time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
	}

	if available && row != nil {
		result.Price = ParseDynadotPrice(row.Price)
		buyURL := AffiliateURL(domain, affiliateTemplate)
		result.BuyURL = &buyURL
	}

	if row != nil && row.Status != "" && row.Status != "success" {
		result.Error = row.Status
	}

	return result
}

func ensureSuccess(payload *dynadotSearchResponse) error {
	if payload.Response != nil && payload.Response.ResponseCode == "-1" {
		msg := payload.Response.Error
		if msg == "" {
			msg = "Dynadot authentication failed"
		}
		if strings.Contains(strings.ToLower(msg), "invalid key") {
			return errors.New("Invalid Dynadot key.")
		}

		return errors.New(msg)
	}

	if payload.SearchResponse != nil && payload.SearchResponse.ResponseCode != "" && payload.SearchResponse.ResponseCode != "0" {
		msg := payload.SearchResponse.Error
		if msg == "" {
			msg = "Dynadot search failed"
		}
		if strings.Contains(strings.ToLower(msg), "invalid key") {
			return errors.New("Invalid Dynadot key.")
		}

		return errors.New(msg)
	}

	return nil
}

type SearchParams struct {
	APIKey            string
	Domains           []string
	Currency          string
	TimeoutMs         int
	AffiliateTemplate string
}

func SearchDynadot(params SearchParams) ([]shared.SearchResult, error) {
	var results []shared.SearchResult

	for _, domain := range params.Domains {
		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(params.TimeoutMs)*time.Millisecond)

		payload, err := requestOne(ctx, params.APIKey, domain, params.Currency)
		cancel()
		if err != nil {
			return nil, err
		}

		if err := ensureSuccess(payload); err != nil {
			return nil, err
		}

		results = append(results, mapResult(domain, payload, params.Currency, params.AffiliateTemplate))
	}

	return results, nil
}
